### PR TITLE
[WIP] wayland-rs-0.20.0

### PR DESCRIPTION
**Pull request by vberger**
Currently WIP, complete rework of the libs to significantly improve the APIs and reduce the unsafe code use, as well as pave the road for a full-rust implementation of the protocol.

**TODO:**

- **commons:**
  - [x] API documentation
  - [x] Crate documentation

- **scanner:**
  - [x] client code generation
  - [x] server code generation

- **client:**
  - [x] object implementation & dispatching
  - [x] connection initialization
  - [x] multiple queues handling (non-send support)
  - [x] convenience code for handling globals
  - [x] wayland-egl
  - [x] wayland-cursor
  - [x] API documentation
  - [x] Crate documentation

- **server:**
  - [x] object implementation & dispatching
  - [x] non-send implementation support
  - [x] server initialization
  - [x] client tracking
  - [x] multiple event queues & event sources
  - [x] API documentation
  - [x] Crate documentation

### CODE REVIEW

**Comment by Timidger** [comment]
> ... as well as pave the road for a full-rust implementation of the protocol.

I must be out of the loop with this...is there still an attempt to rewrite Wayland in Rust?

**Comment by vberger** [comment]
Well, it's mostly a "leave the door open" thing. It's something I've been wanting to work on for some time now, and this refactor allows me to cleanly keep the possibility to do this in the future.

Basically, the idea is that as newer versions of mesa use the dmabuf extension protocol internally (rather than wl_drm), there is technically no long-term blocker to have a full rust implementation of the protocol, at least server-side (I don't know how things are regarding nvidia, but I won't make my hopes high...).

On the other hand, I want to also experiment with ideas such as "rust-implementation of the protocol with a public C api", which would allow wayland-rs to be a system-wide implementation of wayland.

So with this new refactor, the wayland-rs crates have "native_lib" cargo feature, enabled by default, which makes them use the C libs as backend. On day, I may succeed in making a native implementation, in which case removing this feature will use it. (For now, all the code will remain as `unimplemented!()`).

But there is not timeline nor priority for this. The core reason of this refactor is to address usability issues we encountered in smithay, and try to make the API more rust-friendly overall.

**Comment by Timidger** [comment]
Ah ok thanks for the update.

I remember there was talk a while back about skylane being a potential implementation, but was shot down due to the mesa issues you mentioned. The fact that is unblocked is news to me, and very welcome.

I'm way too busy at the moment with Way Cooler and wlroots-rs, but maybe in the future if that is still planned / being worked on I'll look into helping out. At the very least, it will be another implementation which will be nice since everyone just uses the reference one and there should be alternatives (especially in such a safe language).

File: wayland-client/src/lib.rs [hunk 1]
```diff
@@ -1,3 +1,125 @@
+//! Client-side Wayland connector
+//!
+//! ## Overview
+//!
+//! This crate provides the interfaces and machinnery to safely create
```
Comment (mitchmindtree) [inline wayland-client/src/lib.rs]: Typo: machinnery -> machinery

File: wayland-client/src/lib.rs [hunk 2]
```diff
@@ -1,3 +1,125 @@
...
+//! and their `alive()` method will start to return `false`. Trying to send messages
+//! with them will also fail.
+//!
+//! ### Implementations
+//!
+//! To receive and process messages from the server to you (in wayland context they are
+//! called "events"), you need to provide an `Implementation` for each wayland object
+//! created in the protocol session. Whenever a new protocol object is created, you will
+//! receive a `NewProxy<I>` object. Providing an implementation via its `implement()` method
+//! will turn it into a regular `Rroxy<I>` object.
```
Comment (mitchmindtree) [inline wayland-client/src/lib.rs]: Typo: `Rroxy<I>` -> `Proxy<I>` (I think)

File: wayland-client/src/lib.rs [hunk 3]
```diff
@@ -1,3 +1,125 @@
...
+//! around the `libwayland-client.so` C library.
+//!
+//! The wayland protocol revolves around the creation of various objects
+//! and the exchange of messages associated to these objects. The initial
+//! object is always the `Display`, that you get at initialization of the
+//! connection, exposed by this crate as `Display::connect_to_env()`.
+//!
+//! ## Protocol and messages handling model
+//!
+//! The protocol being be-directionnal, you can send and receive messages.
```
Comment (Drakulix) [inline wayland-client/src/lib.rs]: Typo: b**i**-directio**n**al

**Comment by vberger** [comment]
@mitchmindtree @Drakulix : Thanks for the proofreading!

File: wayland-client/examples/dynamic_globals.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    let _globals = GlobalManager::new_with_cb(
+        display.get_registry().unwrap(),
+        // This macro generates a callback for auto-creating the globals
+        // that interest us and calling our provided callbacks
+        global_filter!(
+            // Here we ask that all seats be automatically instanciated
+            // with version 1 when advertized, and provide a callback that
+            // will handle the created wl_seat to implement them
+            //
+            // NOTE: the type annotations are necessayr because rustc's
```
Comment (Drakulix) [inline wayland-client/examples/dynamic_globals.rs]: necessa**ry**

File: wayland-client/examples/dynamic_globals.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+            // inference is apparently not smart enough
+            [
+                wl_seat::WlSeat,
+                1,
+                |seat: Result<NewProxy<_>, _>, ()| {
+                    // here seat is a result, as failure can happen if the server
+                    // advertized an lower version than we requested.
+                    // This cannot happer here as we requested version 1
```
Comment (Drakulix) [inline wayland-client/examples/dynamic_globals.rs]: happe**n**

File: wayland-client/examples/list_globals.rs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+extern crate wayland_client;
+
+use wayland_client::{Display, GlobalManager};
+
+use wayland_client::protocol::wl_display::RequestsTrait;
+
+// A minimal example printing the list of globals abvertized by the server and
```
Comment (Drakulix) [inline wayland-client/examples/list_globals.rs]: a**d**vertized

**Comment by vberger** [comment]
Alright, this has been polished enough and I'm happy with the current state of things. Minor additional fixes shall come in new PRs an releases.

**Pull request merged**